pkg/driver/deployment_awslambda.go

```diff
@@ -3,31 +3,110 @@ package driver
 import (
 	log "github.com/sirupsen/logrus"
 	"github.com/vhive-serverless/loader/pkg/common"
+	"strings"
+	"sync"
+	"sync/atomic"
 )
 
 func DeployFunctionsAWSLambda(functions []*common.Function) {
 	provider := "aws"
 
-	// Create serverless.yml file
-	serverless := Serverless{}
-	serverless.CreateHeader(provider)
-	serverless.AddPackagePattern("./pkg/server/trace-func-go/aws/**")
+	functionGroups := separateFunctions(functions)
 
-	for i := 0; i < len(functions); i++ {
-		serverless.AddFunctionConfig(functions[i], provider)
+	// Use goroutines to create multiple serverless.yml files, deploy functions in parallel, and ensure all finishes
+	// However, due to CPU and memory constraints, we will only deploy 2 serverless.yml files in parallel and wait for them to finish before deploying the next 2
+	var wg sync.WaitGroup
+	var counter uint64 = 0
+	parallelDeployment := 2
+
+	for i := 0; i < len(functionGroups); {
+		for parallelIndex := 0; parallelIndex < parallelDeployment; parallelIndex++ {
+			if i < len(functionGroups) {
+				wg.Add(1)
+				go func(functionGroup []*common.Function, index int) {
+					defer wg.Done()
+					log.Debugf("Creating serverless-%d.yml", index)
+
+					// Create serverless.yml file
+					serverless := Serverless{}
+					serverless.CreateHeader(index, provider)
+					serverless.AddPackagePattern("!**")
+					serverless.AddPackagePattern("./server/trace-func-go/aws/trace_func")
+
+					for i := 0; i < len(functionGroup); i++ {
+						serverless.AddFunctionConfig(functionGroup[i], provider)
+					}
+
+					serverless.CreateServerlessConfigFile(index)
+
+					log.Debugf("Deploying serverless-%d.yml", index)
+					// Deploy serverless functions and update the function endpoints
+					functionToURLMapping := DeployServerless(index)
+
+					if functionToURLMapping == nil {
+						log.Fatalf("Failed to deploy serverless-%d.yml", index)
```

Comment:
Do you need to clean up functions that have successfully been deployed? Does the user get billed for this somehow?

Reply:
Idle functions are not billed, only the invocations and execution time. But the clean up should be done, remaining functions will conflict with the newly deployed.

@LeeXuanHua can you look at this? Can we do a clean up here, so we are cleaning the deployed functions even in the case of failure to deploy some of them?

Reply:
The point is that it is not ran in case of failure to deploy: it exits straight away on Fatalf. But for now, it happens rarely, and we will have separate PR for automatic clean-up and handling such cases.

Reply:
Apologies, you are correct